### PR TITLE
HAL_ChibiOS: prevent long timeouts in DShot

### PR DESCRIPTION
**Pull request by andyp1per**
This is a working version of https://github.com/ArduPilot/ardupilot/pull/20427

Flight tested on Chimera7 bdshot (Lumenier H743)
Tested on MatekH743 with regular and bdshot using assertions - no assertions

Tested on SPRacingH7 with assertions

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -253,9 +255,11 @@ void RCOutput::dshot_collect_dma_locks(uint32_t time_out_us)
             // if we have time left wait for the event
             eventmask_t mask = 0;
             const uint32_t pulse_elapsed_us = now - group.last_dmar_send_us;
-            if (now < time_out_us) {
+            // will be >> timeout_period_us if timeout_us is in the past
+            const uint32_t timeout_remaining_us = timeout_us - now;
+            if (timeout_remaining_us > 0 && timeout_remaining_us <= timeout_period_us) { // wrap sensitive
                 mask = chEvtWaitOneTimeout(group.dshot_event_mask,
-                    chTimeUS2I(MAX(time_out_us - now, group.dshot_pulse_send_time_us - pulse_elapsed_us)));
+                    chTimeUS2I(MAX(timeout_us - now, group.dshot_pulse_send_time_us - pulse_elapsed_us)));
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:262]: I'd like a MIN() with a known small number to protect against any possibility of wrap. 
same goes for the one in the next if()

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Ok added

**Comment by andyp1per** [comment]
Adding the MIN breaks the SPRacing setup - I'll have to dig deeper tomorrow

**Comment by andyp1per** [comment]
There were actually far more wrapping issues. I think these are all nailed now - tested as above on all  the hardware combinations I can think of.

**Comment by tridge** [comment]
@andyp1per can this be closed now, anything you want to keep in it?

**Comment by andyp1per** [comment]
@tridge depends on whether we think the wrapping issues actually matter. I suppose would have to fly past 71minutes to fine out.

**Comment by tridge** [comment]
can we try a fixed timeout?

**Pull request closed without merging**